pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -131,6 +147,29 @@ func (meta *predicateMetadata) RemovePod(deletedPod *v1.Pod) error {
 	}
 	// Delete any anti-affinity rule from the deletedPod.
 	delete(meta.matchingAntiAffinityTerms, deletedPodFullName)
+	// Delete pod from the matching affinity or anti-affinity pods if exists.
+	affinity := meta.pod.Spec.Affinity
+	podNodeName := deletedPod.Spec.NodeName
+	if affinity != nil && len(podNodeName) > 0 {
+		if affinity.PodAffinity != nil {
+			for i, p := range meta.nodeNameToMatchingAffinityPods[podNodeName] {
+				if p == deletedPod {
+					meta.nodeNameToMatchingAffinityPods[podNodeName] = append(meta.nodeNameToMatchingAffinityPods[podNodeName][:i],
+						meta.nodeNameToMatchingAffinityPods[podNodeName][i+1:]...)
```

Comment:
nit: Since this PR is performance related and we are using pointers to pod structs in array, the append may cause potential memory leaks.  I think if we can set the value of the item to nil, it would be better.

Reply:
I am not sure if I understand the concern. This way of deletion of the pod from the array, causes the pod pointer to be removed from the array. When the reference to the pod is removed from the array and other objects, garbage collector should be able to release the memory.

Reply:
While I am not 100% sure IIUC, the pod pointer element will have values for pod.Spec and other items in the struct v1.Pod which won't be deleted by deleting item here. So, in a way they are still referenced by this slice and probably not garbage collected. I think, it is always recommended in go if the elements are pointers is to copy and create a nil value for it.

Ref: https://github.com/golang/go/wiki/SliceTricks

Reply:
> Ref: https://github.com/golang/go/wiki/SliceTricks

Oh, that's a slightly different problem. When you shrink an array length (like we do in delete), but not the array capacity, a few items at the end of the array (between array length and array capacity) remain in memory. When the items are pointers, those pointers keep referencing objects and prevent garbage collector from cleaning those objects. So, it is recommended to set those items to nil. Those items are not the deleted items.
In this piece of code, we should be fine as the whole array is deleted at the end of a scheduling cycle and all of its elements will go.

Reply:
Thanks Bobby for explanation.

> So, it is recommended to set those items to nil. Those items are not necessarily the deleted items. In this piece of code, we should be fine as the whole array is deleted at the end of a scheduling cycle and all of its elements will go.

To be clear, in delete if we use the append method as above(with pointers), we will have garbage collection issues but since we are deleting this array at the end of scheduling cycle, we won't have problem(?)


> But your link reminded me that we should swap the last element of the array with the deleted item and shrink the array length instead of removing an item from the middle of array. That method would be faster and avoids copying many elements.

Right. Theoretically yes. But last time I looked around, the benchmark differences were in range of 6-10 ns/op, so I am not sure if it would add much of benefit(To be fair, the data type was a int and it was in go 1.7, so we may have some improvement).